src/nni_engine.hpp

```diff
@@ -57,47 +63,79 @@ class NNIEngine {
   // Get number of runs of NNI engine.
   size_t GetSweepCount() const { return sweep_count_; };
 
+  // Set GP Engine.
+  void SetGPEngine(GPEngine *gp_engine) { gp_engine_ = gp_engine; }
+  // Set TP Engine.
+  void SetTPEngine(TPEngine *tp_engine) { tp_engine_ = tp_engine; }
+
   // ** Runners
   // These start the engine, which procedurally ranks and adds (and maybe removes) NNIs
   // to the DAG, until some termination criteria has been satisfied.
 
-  // Primary Runner using GraftDAG.
+  // Primary Runner for NNI Engine.
   void Run();
-  // Runner subroutines.
+  // Initialization step run before loop.
   void RunInit();
+  // Step that finds adjacent NNIs, evaluates, tehn accepts or rejects them.
   void RunMainLoop();
+  // Step that runs at the end of each loop, preps for next loop.
   void RunPostLoop();
 
   // ** Filter Functions
 
   // Initialization step for filter before beginning
   using StaticFilterInitFunction =
-      std::function<void(NNIEngine &, GPEngine &, GraftDAG &)>;
+      std::function<void(NNIEngine &, GPEngine &, TPEngine &, GraftDAG &)>;
```

Comment:
That is a good point.  I think we should add declare a some common GP/TP/XYEngine API interface, then NNI Engine can hold one or more of those engines and rather than NNI Engine storing any maintenance logic, can just make direct calls to this interface.

Reply:
Great. Keep that thought going but let's stay the course for now.